concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -479,87 +480,116 @@ doStoreBlock pb@GB.PendingBlock{..} = unlessShutDown $ do
+doReceiveTransactionInternal :: (TreeStateMonad pv m) => TransactionOrigin -> BlockItem -> Timestamp -> Slot -> m (Maybe BlockItem, (UpdateResult, TransactionReceiveOutcome))
+doReceiveTransactionInternal origin tr ts slot = do
+   -- Don't accept the transaction if it was expired
+   if transactionExpired (msgExpiry tr) ts then return (Nothing, (ResultStale, Rejected))
+    else do
+      bs <- blockState =<< getFocusBlock
+      gd <- getGenesisData
+      verRes <- runReaderT (TV.verify ts tr) $ TV.Context bs (gdMaxBlockEnergy gd)
+      if definitelyNotValid verRes then return (Nothing, (mapTransactionVerificationResult verRes, Rejected))
+      else
+        addTx verRes
+  where
+      addTx verRes = addCommitTransaction (tr, verRes) slot >>= \case
           Added bi@WithMetadata{..} -> do
-              ptrs <- getPendingTransactions
-              case wmdData of
-                NormalTransaction tx -> do
-                  focus <- getFocusBlock
-                  st <- blockState focus
-                  macct <- getAccount st $! transactionSender tx
-                  nextNonce <- fromMaybe minNonce <$> mapM (getAccountNonce . snd) macct
-                  -- If a transaction with this nonce has already been run by
-                  -- the focus block, then we do not need to add it to the
-                  -- pending transactions.  Otherwise, we do.
-                  when (nextNonce <= transactionNonce tx) $
-                      putPendingTransactions $! addPendingTransaction nextNonce WithMetadata{wmdData=tx,..} ptrs
-                CredentialDeployment _ ->
-                  putPendingTransactions $! addPendingDeployCredential wmdHash ptrs
-                ChainUpdate cu -> do
-                    focus <- getFocusBlock
-                    st <- blockState focus
-                    nextSN <- getNextUpdateSequenceNumber st (updateType (uiPayload cu))
-                    when (nextSN <= updateSeqNumber (uiHeader cu)) $
-                        putPendingTransactions $! addPendingUpdate nextSN cu ptrs
-              return (Just bi, ResultSuccess)
-          Duplicate tx -> return (Just tx, ResultDuplicate)
-          ObsoleteNonce -> return (Nothing, ResultStale)
+            ptrs <- getPendingTransactions
+            case wmdData of
+              NormalTransaction tx -> do
+                putPendingTransactions $! addPendingTransaction (transactionNonce tx) WithMetadata{wmdData=tx,..} ptrs
+              CredentialDeployment _ -> do
+                putPendingTransactions $! addPendingDeployCredential wmdHash ptrs
+              ChainUpdate cu -> do
+                putPendingTransactions $! addPendingUpdate (updateSeqNumber (uiHeader cu)) cu ptrs
+            -- The actual verification result here is only used if the transaction was received individually.
+            -- If the transaction was received as part of a block we don't use the result for anything.
+            -- todo: should we use it for e.g. logging?
```

Comment:
Resolve this todo?

It might be useful to log, but we need to be careful to limit the logging.

Reply:
Decided not to log anything here. 
1. `rpc.rs` will log the single transaction failure.  
2. For transactions received as part of a block, I will add some cautious logging in `doStoreBlock`.